new-tutorial/L2_large_link_prediction.py

```diff
@@ -0,0 +1,361 @@
+# representation of output nodes.
+# 
+# The second element and the third element are the positive graph and the
+# negative graph for this minibatch. Recall that in the :doc:`full-graph link
+# prediction tutorial <4_link_predict>`,
+# 
+# -  The positive graph contains all the nodes in the graph and all the
+#    edges in the training set.
+# -  The negative graph contains all the nodes in the graph and all the
+#    negative examples generated before training as edges.
+# 
+# In minibatch training, the positive graph and the negative graph only
+# contain all the nodes necessary for computing the positive examples in
+# the current minibatch, as well as their corresponding negative examples
+# generated by ``negative_sampler`` you have defined earlier.
+# 
+# The last element is the list of bipartite graphs, one for each layer.
+# The bipartite graphs are used to compute the GNN outputs of the nodes
+# involved in positive/negative graph.
+# 
+
+
+######################################################################
+# Defining Model for Node Representation
+# --------------------------------------
+# 
+# The model is almost identical to the one in the :doc:`node classification
+# tutorial <L1_large_node_classification>`. The only difference is
+# that since you are doing link prediction, the output dimension will not
+# be the number of classes in the dataset.
+# 
+
+import torch.nn as nn
+import torch.nn.functional as F
+from dgl.nn import SAGEConv
+
+class Model(nn.Module):
+    def __init__(self, in_feats, h_feats):
+        super(Model, self).__init__()
+        self.conv1 = SAGEConv(in_feats, h_feats, aggregator_type='mean')
+        self.conv2 = SAGEConv(h_feats, h_feats, aggregator_type='mean')
+        self.h_feats = h_feats
+        
+    def forward(self, bipartites, x):
+        h_dst = x[:bipartites[0].num_dst_nodes()]
+        h = self.conv1(bipartites[0], (x, h_dst))
+        h = F.relu(h)
+        h_dst = h[:bipartites[1].num_dst_nodes()]
+        h = self.conv2(bipartites[1], (h, h_dst))
+        return h
+    
+model = Model(num_features, 128).cuda()
+
+
+######################################################################
+# Defining the Score Predictor for Edges
+# --------------------------------------
+# 
+# After getting the node representation necessary for the minibatch, the
+# last thing to do is to predict the score of the edges and non-existent
+# edges in the sampled minibatch.
+# 
+# The following score predictor, copied from the :doc:`link prediction
+# tutorial <4_link_predict>`, takes a dot product between the
+# incident nodes’ representations.
+# 
+
+import dgl.function as fn
+
+class DotPredictor(nn.Module):
+    def forward(self, g, h):
+        with g.local_scope():
+            g.ndata['h'] = h
+            g.apply_edges(fn.u_dot_v('h', 'h', 'score'))
+            # u_dot_v returns a 1-element vector for each edge so you need to squeeze it.
+            return g.edata['score'][:, 0]
```

Comment:
if you change anything in the node-class tutorial, remember to copy-paste it here

Reply:
I assume you mean the full graph link prediction tutorial.